### ISSUE TITLE
Facebook embeds don't return an icon compared to the hosted solution

### ISSUE THREAD

**Issue by stubbegianni** [itteco/iframely#172]
Hello,

I'm seeing an issue embedding facebook on my self hosted solution. When I enter the following url `https://www.facebook.com/dansercoer` I get the following response: 
```
{
  "meta": {
    "title": "dansercoer",
    "canonical": "https://www.facebook.com/dansercoer",
    "site": "Facebook"
  },
  "links": {},
  "rel": []
}
```

When I do the same thing on the cloud solution I get: 

```
{
    "url": "https://www.facebook.com/dansercoer",
    "meta": {
        "site": "Facebook",
        "canonical": "https://www.facebook.com/dansercoer",
        "title": "dansercoer"
    },
    "links": {
        "icon": [
            {
                "type": "image/icon",
                "rel": [
                    "shortcut",
                    "icon",
                    "icon",
                    "ssl"
                ],
                "href": "https://static.xx.fbcdn.net/rsrc.php/yV/r/hzMapiNYYpW.ico"
            }
        ]
    },
    "rel": []
}
```

Am I doing something wrong in my configuration as facebook is the only one having this issue. 

my config if needed: 
```
(function() {
    var config = {

        DEBUG: false,
        RICH_LOG_ENABLED: false,

        // For embeds that require render, baseAppUrl will be used as the host.
        baseAppUrl: "http://embed.site.com",
        relativeStaticUrl: "/r",

        // Or just skip built-in renders altogether
        SKIP_IFRAMELY_RENDERS: true,

        // For legacy reasons the response format of Iframely open-source is
        // different by default as it does not group the links array by rel.
        // In order to get the same grouped ressonse as in Cloud API,
        // add `&group=true` to your request to change response per request
        // or set `groupLinks` in your config to `true` for a global change.
        GROUP_LINKS: true,

        // Number of maximum redirects to follow before aborting the page
        // request with `redirect loop` error.
        MAX_REDIRECTS: 4,

        SKIP_OEMBED_RE_LIST: [
            // /^https?:\/\/yourdomain\.com\//,
        ],

        /*
        // Used to pass parameters to the generate functions when creating HTML elements
        // disableSizeWrapper: Don't wrap element (iframe, video, etc) in a positioned div
        GENERATE_LINK_PARAMS: {
            disableSizeWrapper: true
        },
        */

        port: 8061, //can be overridden by PORT env var
        host: '0.0.0.0', //can be overridden by HOST env var

        // Optional SSL cert, if you serve under HTTPS.
        /*
        ssl: {
            key: require('fs').readFileSync(__dirname + '/key.pem'),
            cert: require('fs').readFileSync(__dirname + '/cert.pem'),
            port: 443
        },
        */

        /*
        Supported cache engines:
        - no-cache - no caching will be used.
        - node-cache - good for debug, node memory will be used (https://github.com/tcs-de/nodecache).
        - redis - https://github.com/mranney/node_redis.
        - memcached - https://github.com/3rd-Eden/node-memcached
        */
        CACHE_ENGINE: 'no-cache',
        CACHE_TTL: 0, // In milliseconds. 0 for 'never expire' to let cache engine decide itself when to evict the record

        /*
        // Redis cache options.
        REDIS_OPTIONS: {
            host: '127.0.0.1',
            port: 6379
        },
        */

        /*
        // Memcached options. See https://github.com/3rd-Eden/node-memcached#server-locations
        MEMCACHED_OPTIONS: {
            locations: "127.0.0.1:11211"
        }
        */

        /*
        // Access-Control-Allow-Origin list.
        allowedOrigins: [
            "*",
            "http://another_domain.com"
        ],
        */

        /*
        // Uncomment to enable plugin testing framework.
        tests: {
            mongodb: 'mongodb://localhost:27017/iframely-tests',
            single_test_timeout: 10 * 1000,
            plugin_test_period: 2 * 60 * 60 * 1000,
            relaunch_script_period: 5 * 60 * 1000
        },
        */

        // If there's no response from remote server, the timeout will occur after
        RESPONSE_TIMEOUT: 5 * 1000, //ms

        // Customize API calls to oembed endpoints.
        ADD_OEMBED_PARAMS: [{
            // Endpoint url regexp array.
            re: [/^http:\/\/api\.instagram\.com\/oembed/],
            // Custom get params object.
            params: {
                hidecaption: true
            }
        }, {
            re: [/^https:\/\/www\.facebook\.com\/plugins\/page\/oembed\.json/i],
            params: {
                show_posts: 0,
                show_facepile: 0,
                maxwidth: 600
            }
        }, {
            // match i=user or i=moment or i=timeline to configure these types invidually
            // see params spec at https://dev.twitter.com/web/embedded-timelines/oembed
            re: [/^https?:\/\/publish\.twitter\.com\/oembed\?i=user/i],
            params: {
                limit: 1,
                maxwidth: 600
            }
        }, {
            // Facebook https://developers.facebook.com/docs/plugins/oembed-endpoints
            re: [/^https:\/\/www\.facebook\.com\/plugins\/\w+\/oembed\.json/i],
            params: {
                // Skip script tag and fb-root div.
                omitscript: true
            }
         }],

        /*
        // Configure use of HTTP proxies as needed
        PROXY: [{
            re: [/^https?:\/\/www\.domain\.com/],
            proxy_server: 'http://1.2.3.4:8080',
            // user_agent: 'CHANGE YOUR AGENT'
        }],
        */

        // Customize API calls to 3rd parties. At the very least - configure required keys.
        providerOptions: {
            "twitter": {
                "max-width": 550,
                "min-width": 250,
                hide_media: false,
                hide_thread: false,
                omit_script: false,
                cache_ttl: 100 * 365 * 24 * 3600 // 100 Years.
            },
            readability: {
                enabled: false
                // allowPTagDescription: true  // to enable description fallback to first paragraph
            },
            images: {
                loadSize: false, // if true, will try an load first bytes of all images to get/confirm the sizes
                checkFavicon: false // if true, will verify all favicons
            },
            tumblr: {
                consumer_key: "INSERT YOUR VALUE"
                // media_only: true     // disables status embeds for images and videos - will return plain media
            },
            google: {
                // https://developers.google.com/maps/documentation/embed/guide#api_key
                maps_key: "myapikeyhere"
            },

            /*
            // Optional Camo Proxy to wrap all images: https://github.com/atmos/camo
            camoProxy: {
                camo_proxy_key: "INSERT YOUR VALUE",
                camo_proxy_host: "INSERT YOUR VALUE"
            // Start it with leading "?". Or omit alltogether for default values
            // API key is optional, youtube will work without it too.
            // It is probably the same API key you use for Google Maps.
            youtube: {
                // api_key: "INSERT YOUR VALUE",
                get_params: "?rel=0&showinfo=1"     // https://developers.google.com/youtube/player_parameters
            },
            vimeo: {
                get_params: "?byline=0&badge=0"     // http://developer.vimeo.com/player/embedding
            },

            /*
            soundcloud: {
                old_player: true // enables classic player
            },
            giphy: {
                media_only: true // disables branded player for gifs and returns just the image
            }
            */
            /*
            bandcamp: {
                get_params: '/size=large/bgcol=333333/linkcol=ffffff/artwork=small/transparent=true/',
                media: {
                    album: {
                        height: 472,
                        'max-width': 700
                    },
                    track: {
                        height: 120,
                        'max-width': 700
                    }
                }
            }
            */
        },

        // WHITELIST_WILDCARD, if present, will be added to whitelist as record for top level domain: "*"
        // with it, you can define what parsers do when they run accross unknown publisher.
        // If absent or empty, all generic media parsers will be disabled except for known domains
        // More about format: https://iframely.com/docs/qa-format

        /*
        WHITELIST_WILDCARD: {
              "twitter": {
                "player": "allow",
                "photo": "deny"
              },
              "oembed": {
                "video": "allow",
                "photo": "allow",
                "rich": "deny",
                "link": "deny"
                "link": "deny"
              },
              "og": {
                "video": ["allow", "ssl", "responsive"]
              },
              "iframely": {
                "survey": "allow",
                "reader": "allow",
                "player": "allow",
                "image": "allow"
              },
              "html-meta": {
                "video": ["allow", "responsive"],
                "promo": "allow"
              }
        }
        */
    };

    module.exports = config;
})();

```

**Comment by iparamonau** [comment]
Please try it with `providerOptions.images.checkFavicon: true`. 

The other thing is that the way FB parser works with favicons is that the first request after the fresh server restart will not return a favicon (but will prefetch in the background), and the next requests will get the favicons data from the cache.

**Comment by iparamonau** [comment]
Actually, your config is for `CACHE_ENGINE: 'no-cache'`. That way, yes - even subsequent requests will have no favicons, as they are only returned from the cache for FB.

**Comment by stubbegianni** [comment]
Hello,

I have tried to add the provider options and added node-cache to test but I still receive the same result. 
The cache is working as it logs it that it is using it.

I was running node 8 and tried it with node 7 but I'm getting the same result over there as well, still no icon there. 

Any further advice?

**Comment by iparamonau** [comment]
You could had run into the issue with the redirects on the root URL for the domain. Perhaps, Facebook is redirecting it's main page when requesting server IP is outside of US. 

Commit above should fix that, will be merged to master for the next release. Thanks for asking!

**Comment by stubbegianni** [comment]
Ok I tried out with the latest version on github and I'm still seeing the issue but I have more exact info.

When I request something like `http://imgur.com/a/z72ln` I get no icon at first for imgur.com But next request for that domain do include it. Only the first request will have been cached and will have no icon for any subsequent request on that specific url. Any idea how to circumvent this issue? 

So in short:
Request one (no cache) => http://.../iframely?url=http://imgur.com/a/kpjzY => no icon
request two (with cache) => http://.../iframely?url=http://imgur.com/a/z72ln => icon returned
request three (with cache) => http://.../iframely?url=http://imgur.com/a/kpjzY => still no icon as this request itself has been cached I suppose

**Comment by iparamonau** [comment]
That is the designed behaviour for performance reasons. Number of domain plugins use `domain-icon` mixin that works the way you described. You can change it locally. Just copy `custom\domain-icon.js` file into your `CUSTOM_PLUGINS_PATH` folder (see config example) and modify the mixin the way you like.
